### PR TITLE
JIRA: EDGOAIPMH-15 - Create OAIPMHOkapiClient and OAIPMHOkapiClientFactory

### PR DESCRIPTION
**Pull request by khandramai**
Client and ClientFactory implemented
Viachaslau Khandramai <Viachaslau_Khandramai@epam.com>

### CODE REVIEW

File: .idea/encodings.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (craigmcnally) [inline .idea/encodings.xml]: Is this file specific to your IDE/dev env?  If so, I don't think it needs to be checked-in

File: .idea/sonarIssues.xml [hunk 2]
```diff
@@ -0,0 +1,214 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (craigmcnally) [inline .idea/sonarIssues.xml]: Same as above... I'm not sure this needs to be checked-in

File: .idea/encodings.xml (hunk 1 above)
Comment (epam-pavelk) [reply .idea/encodings.xml]: This is a folder created by Intellij IDEA. I would propose to add it to .gitignore of the module similar to Eclipse specific folders/files.
@craigmcnally is it acceptable?

File: src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java [hunk 3]
```diff
@@ -2,23 +2,133 @@
...
+    return endpoint;
+  }
+
+  /**
+   * This method transform MultiMap of HTTP GET request parameters to string representation
+   *
+   * @param parameters HTTP GET parameters multimap
+   * @return string representation of GET request parameters
+   */
+  private String getParametersAsString(MultiMap parameters) {
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java]: I'm pretty sure that you can get this via the VertX RoutingContext object passed to the route handler(s).  E.g. routingCtx.request().query().

File: src/test/java/org/folio/edge/oaipmh/utils/OaiPmhMockOkapi.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        .end(getOaiPmhResponseAsXml(Paths.get(PATH_TO_IDENTIFY_MOCK)));
+    }
+  }
+
+  static String getOaiPmhResponseAsXml(Path pathToXmlFile) {
+    String xml = null;
+    try {
+      xml = new String(Files.readAllBytes(pathToXmlFile));
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/oaipmh/utils/OaiPmhMockOkapi.java]: Please use a logger

File: src/test/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClientFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import io.vertx.core.Vertx;
+import org.junit.Before;
+import org.junit.Test;
+
+public class OaiPmhOkapiClientFactoryTest {
+
+  private static final long reqTimeout = 5000L;
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClientFactoryTest.java]: Use the requestTimeout provided via SystemProperties and ready by EdgeVerticle.

File: src/test/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClientFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+  private OaiPmhOkapiClientFactory ocf;
+
+  @Before
+  public void setUp() {
+    Vertx vertx = Vertx.vertx();
+    ocf = new OaiPmhOkapiClientFactory(vertx, "http://mocked.okapi:9130", reqTimeout);
```
Comment (craigmcnally) [inline src/test/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClientFactoryTest.java]: Use the OkapiURL provided via System Properties and read by EdgeVerticle

File: .idea/encodings.xml (hunk 1 above)
Comment (craigmcnally) [reply .idea/encodings.xml]: yes, sounds good

File: src/test/java/org/folio/edge/oaipmh/utils/OaiPmhMockOkapi.java (hunk 4 above)
Comment (khandramai) [reply src/test/java/org/folio/edge/oaipmh/utils/OaiPmhMockOkapi.java]: Fixed. Thanks.

File: .idea/sonarIssues.xml (hunk 2 above)
Comment (khandramai) [reply .idea/sonarIssues.xml]: Fixed. Thanks.

File: .idea/encodings.xml (hunk 1 above)
Comment (khandramai) [reply .idea/encodings.xml]: Fixed. Thanks.

File: src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java (hunk 3 above)
Comment (craigmcnally) [reply src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java]: > ah, good point @epam-pavelk

File: src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java [hunk 7]
```diff
@@ -129,6 +128,6 @@ private String getParametersAsString(MultiMap parameters) {
    * @return URL for corresponding 'verb'
    */
   private String getUrlByVerb(MultiMap parameters) {
-    return okapiURL + getEndpoint(parameters) + getParametersAsString(parameters);
+    return okapiURL + getEndpoint(parameters) + "?" + getParametersAsString(parameters);
```
Comment (craigmcnally) [inline src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java]: not a big deal, but in the future consider using String.format(...)

Comment (khandramai) [reply src/main/java/org/folio/edge/oaipmh/utils/OaiPmhOkapiClient.java]: @craigmcnally,
noted. Thanks.

**Pull request merged**